libraries/fc/src/network/udp_socket.cpp

```diff
@@ -1,166 +1,105 @@
 #include <fc/network/udp_socket.hpp>
 #include <fc/network/ip.hpp>
 #include <fc/fwd_impl.hpp>
-#include <fc/asio.hpp>
 
-
-namespace fc {
-  using boost::fibers::promise;
-  
-  class udp_socket::impl {
+namespace fc
+{
+  class udp_socket::impl : public fc::retainable
+  {
     public:
-      impl():_sock( fc::asio::default_io_service() ){}
-      ~impl(){ }
+      impl(){}
+      ~impl(){}
 
-      boost::asio::ip::udp::socket _sock;
+      std::shared_ptr<boost::asio::ip::udp::socket> _sock;
   };
 
-  boost::asio::ip::udp::endpoint to_asio_ep( const fc::ip::endpoint& e ) {
-    return boost::asio::ip::udp::endpoint(boost::asio::ip::address_v4(e.get_address()), e.port() );
-  }
-  fc::ip::endpoint to_fc_ep( const boost::asio::ip::udp::endpoint& e ) {
-    return fc::ip::endpoint( e.address().to_v4().to_ulong(), e.port() );
+  udp_socket::udp_socket()
+    : my(new impl())
+  {
   }
 
-  udp_socket::udp_socket()
-  :my( new impl() ) 
+  void udp_socket::initialize(boost::asio::io_service& service)
   {
+    my->_sock.reset(new boost::asio::ip::udp::socket(service));
   }
 
   udp_socket::~udp_socket() 
   {
     try 
     {
-      my->_sock.close(); //close boost socket to make any pending reads run their completion handler
+      if(my->_sock)
+        my->_sock->close(); //close boost socket to make any pending reads run their completion handler
     }
     catch (...) //avoid destructor throw and likely this is just happening because socket wasn't open.
     {
     }
   }
 
-  size_t udp_socket::send_to( const char* buffer, size_t length, const ip::endpoint& to ) 
+  void udp_socket::send_to(const char* buffer, size_t length, boost::asio::ip::udp::endpoint& to)
   {
     try 
     {
-      return my->_sock.send_to( boost::asio::buffer(buffer, length), to_asio_ep(to) );
+      my->_sock->send_to(boost::asio::buffer(buffer, length), to);
     } 
-    catch( const boost::system::system_error& e ) 
+    catch(const boost::system::system_error& e)
     {
-      if( e.code() != boost::asio::error::would_block )
+      if(e.code() != boost::asio::error::would_block)
         throw;
     }
 
-    auto completion_promise = std::make_shared<promise<size_t>>();
-    my->_sock.async_send_to( boost::asio::buffer(buffer, length), to_asio_ep(to), 
-                             asio::detail::read_write_handler(completion_promise) );
-
-    return completion_promise->get_future().get();
-  }
-
-  size_t udp_socket::send_to( const std::shared_ptr<const char>& buffer, size_t length, 
-                              const fc::ip::endpoint& to )
-  {
-    try 
+    my->_sock->async_send_to(boost::asio::buffer(buffer, length), to,
+                             [this](const boost::system::error_code& /*ec*/, std::size_t /*bytes_transferred*/)
```

Comment:
Should not capture `this`. It is not used. 
Also your `buffer` is not protected from going out of scope. The previous code used `completion_promise` to make sure buffer was valid for life of `async_sent_to`.

Reply:
Fixed.